File: java/r2dbc-backend-2/src/main/java/chaincue/tech/r2dbcbackend2/masters/DomainObject.java
package chaincue.tech.r2dbcbackend2.masters;

import org.springframework.data.annotation.Transient;
import org.springframework.data.domain.Persistable;

public interface DomainObject extends Persistable<String> {

    default boolean createNewDomain() {
        return true;
    }

    @Override
    @Transient
    default boolean isNew() {
        return createNewDomain() || getId() == null;
    }
}

File: kotlin/chaincue-real-estate-reactive/src/main/kotlin/com/chaincuerealestate/chaincuerealestatereactive/domains/DomainObject.kt
package com.chaincuerealestate.chaincuerealestatereactive.domains

import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Transient
import org.springframework.data.domain.Persistable
import java.util.*

abstract class DomainObject(
    @Id private val id: String = UUID.randomUUID().toString(),
    @Transient private val isNew: Boolean = true
) : Persistable<String> {
    override fun getId() = id
    @JsonIgnore
    override fun isNew() = isNew
}
